File: cursor_tolerance_test.go
package yaylib

import (
	"encoding/json"
	"testing"

	"github.com/ekkx/yaylib/v2/gen"
)

// The pagination cursor is a string in the schema, but the server
// sends it as a bare JSON number on some endpoints and a JSON string
// on others. Typed decoding must accept both and expose the string
// form either way.
func TestPaginationCursorToleratesNumberAndString(t *testing.T) {
	cases := []struct{ name, body, want string }{
		{"number wire", `{"next_page_value":585851614,"posts":[]}`, "585851614"},
		{"string wire", `{"next_page_value":"abc123","posts":[]}`, "abc123"},
		{"null wire", `{"next_page_value":null,"posts":[]}`, ""},
	}
	for _, c := range cases {
		var r gen.PostsResponse
		if err := json.Unmarshal([]byte(c.body), &r); err != nil {
			t.Fatalf("%s: unmarshal failed: %v", c.name, err)
		}
		if got := r.GetNextPageValue(); got != c.want {
			t.Fatalf("%s: GetNextPageValue() = %q, want %q", c.name, got, c.want)
		}
	}
}

File: packages/typescript/tests/cursor_tolerance.test.ts
// The pagination cursor is a string in the schema, but the server
// sends it as a bare JSON number on some endpoints and a JSON string
// on others. FromJSON must accept both and yield the string form.

import { PostsResponseFromJSON } from "../src/gen/models/PostsResponse";

let failed = 0;
function assert(name: string, cond: boolean, detail = ""): void {
  if (cond) console.log(`PASS ${name}`);
  else {
    failed++;
    console.log(`FAIL ${name}${detail ? `\n  ${detail}` : ""}`);
  }
}

const num = PostsResponseFromJSON({ next_page_value: 585851614, posts: [] });
assert(
  "number wire -> string",
  num.nextPageValue === "585851614",
  `got ${JSON.stringify(num.nextPageValue)}`,
);

const str = PostsResponseFromJSON({ next_page_value: "abc123", posts: [] });
assert(
  "string wire -> string",
  str.nextPageValue === "abc123",
  `got ${JSON.stringify(str.nextPageValue)}`,
);

const nul = PostsResponseFromJSON({ next_page_value: null, posts: [] });
assert(
  "null wire -> undefined",
  nul.nextPageValue === undefined,
  `got ${JSON.stringify(nul.nextPageValue)}`,
);

process.exit(failed === 0 ? 0 : 1);
